### ISSUE TITLE
RocksDB Immutable keys prototype

### ISSUE THREAD

**Issue by gz** [feldera/feldera#777]
- Build a rocksdb storage prototype that stores everything except weight as keys
- Evaluate it by running a subset of (slow) nexmark queries

**Comment by gz** [comment]
tldr is that this design with rocksdb is unlikely to be a better choice than what we currently have in main -- so closing as wont fix

the branch is here:
https://github.com/feldera/feldera/tree/immutable-keys

note that 
https://github.com/feldera/feldera/commit/197018402f118e93768e79474a26ff5ed3a69b87
and
https://github.com/feldera/feldera/commit/c28429a142c367c5b93d30868def6d75ae67f8ed

will be useful in case we use a non-rocksdb solution because it enables zero-cost de-serialization with rkyv


MAIN BRANCH (where a dbsp key is a rocksdb key, and dbsp values are in a rocksdb value as a vector):
┌───────┬───────────┬───────┬───────────┬─────────────────┬──────────────────┬───────────────┬───────────────┬─────────────┬─────────────┬────────────────┬─────────────┬─────────────┐
│ Query │ #Events   │ Cores │ Elapsed   │ Cores * Elapsed │ Throughput/Cores │ Total Usr CPU │ Total Sys CPU │ Current RSS │ Peak RSS    │ Current Commit │ Peak Commit │ Page Faults │
├───────┼───────────┼───────┼───────────┼─────────────────┼──────────────────┼───────────────┼───────────────┼─────────────┼─────────────┼────────────────┼─────────────┼─────────────┤
│ q0    │ 3,000,000 │ 1     │ 605.816ms │ 605.816ms       │ 4951.996 K/s     │ 3.553s        │ 64.000ms      │ 143.17 MiB  │ 143.27 MiB  │ 184.73 MiB     │ 184.83 MiB  │ 0           │
│ q1    │ 3,000,000 │ 1     │ 630.393ms │ 630.393ms       │ 4758.936 K/s     │ 3.585s        │ 48.000ms      │ 171.11 MiB  │ 171.36 MiB  │ 16.92 MiB      │ 17.02 MiB   │ 0           │
│ q2    │ 3,000,000 │ 1     │ 606.975ms │ 606.975ms       │ 4942.544 K/s     │ 3.452s        │ 45.000ms      │ 176.64 MiB  │ 176.92 MiB  │ 944.00 KiB     │ 1.02 MiB    │ 0           │
│ q3    │ 3,000,000 │ 1     │ 614.708ms │ 614.708ms       │ 4880.363 K/s     │ 3.508s        │ 50.000ms      │ 193.91 MiB  │ 193.98 MiB  │ 944.00 KiB     │ 1.02 MiB    │ 0           │
│ q4    │ 3,000,000 │ 1     │ 2.140s    │ 2.140s          │ 1401.835 K/s     │ 5.067s        │ 140.000ms     │ 365.23 MiB  │ 365.31 MiB  │ 8.94 MiB       │ 9.03 MiB    │ 0           │
│ q5    │ 3,000,000 │ 1     │ 627.791ms │ 627.791ms       │ 4778.663 K/s     │ 3.491s        │ 54.000ms      │ 368.59 MiB  │ 368.67 MiB  │ 944.00 KiB     │ 1.02 MiB    │ 0           │
│ q6    │ 3,000,000 │ 1     │ 2.314s    │ 2.314s          │ 1296.444 K/s     │ 5.264s        │ 109.000ms     │ 519.55 MiB  │ 519.62 MiB  │ 944.00 KiB     │ 1.02 MiB    │ 0           │
│ q7    │ 3,000,000 │ 1     │ 1.269s    │ 1.269s          │ 2363.489 K/s     │ 4.135s        │ 186.000ms     │ 589.17 MiB  │ 589.33 MiB  │ 16.94 MiB      │ 17.02 MiB   │ 0           │
│ q8    │ 3,000,000 │ 1     │ 613.388ms │ 613.388ms       │ 4890.872 K/s     │ 3.462s        │ 45.000ms      │ 590.11 MiB  │ 590.42 MiB  │ 944.00 KiB     │ 1.02 MiB    │ 0           │
│ q9    │ 3,000,000 │ 1     │ 5.756s    │ 5.756s          │ 521.166 K/s      │ 9.948s        │ 667.000ms     │ 954.31 MiB  │ 954.39 MiB  │ 272.92 MiB     │ 273.02 MiB  │ 0           │
│ q12   │ 3,000,000 │ 1     │ 1.055s    │ 1.055s          │ 2844.729 K/s     │ 3.964s        │ 69.000ms      │ 972.31 MiB  │ 972.45 MiB  │ 944.00 KiB     │ 1.02 MiB    │ 0           │
│ q13   │ 3,000,000 │ 1     │ 2.440s    │ 2.440s          │ 1229.310 K/s     │ 6.798s        │ 166.000ms     │ 1002.08 MiB │ 1002.41 MiB │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q14   │ 3,000,000 │ 1     │ 615.470ms │ 615.470ms       │ 4874.326 K/s     │ 4.212s        │ 51.000ms      │ 1004.23 MiB │ 1004.55 MiB │ 16.00 EiB      │ 128.00 KiB  │ 0           │
│ q15   │ 3,000,000 │ 1     │ 2.362s    │ 2.362s          │ 1270.150 K/s     │ 5.364s        │ 157.000ms     │ 936.67 MiB  │ 1012.44 MiB │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q16   │ 3,000,000 │ 1     │ 12.257s   │ 12.257s         │ 244.762 K/s      │ 14.863s       │ 327.000ms     │ 920.36 MiB  │ 1012.44 MiB │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q17   │ 3,000,000 │ 1     │ 6.633s    │ 6.633s          │ 452.303 K/s      │ 9.348s        │ 267.000ms     │ 1.02 GiB    │ 1.02 GiB    │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q18   │ 3,000,000 │ 1     │ 4.796s    │ 4.796s          │ 625.497 K/s      │ 8.814s        │ 560.000ms     │ 1.08 GiB    │ 1.08 GiB    │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q19   │ 3,000,000 │ 1     │ 3.085s    │ 3.085s          │ 972.417 K/s      │ 6.417s        │ 570.000ms     │ 1.11 GiB    │ 1.11 GiB    │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q20   │ 3,000,000 │ 1     │ 1.347s    │ 1.347s          │ 2226.963 K/s     │ 4.662s        │ 370.000ms     │ 1.11 GiB    │ 1.11 GiB    │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q21   │ 3,000,000 │ 1     │ 607.005ms │ 607.005ms       │ 4942.297 K/s     │ 3.540s        │ 45.000ms      │ 1.11 GiB    │ 1.11 GiB    │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q22   │ 3,000,000 │ 1     │ 608.689ms │ 608.689ms       │ 4928.627 K/s     │ 3.518s        │ 45.000ms      │ 1.11 GiB    │ 1.11 GiB    │ 16.00 EiB      │ 144.00 KiB  │ 0           │
└───────┴───────────┴───────┴───────────┴─────────────────┴──────────────────┴───────────────┴───────────────┴─────────────┴─────────────┴────────────────┴─────────────┴─────────────┘
╔════╦══════════════════════════════════════════════════════╦════════════╦═══════════╦══════════╦═════════════╗
║ id ║ span name                                            ║ call count ║ called by ║ total    ║ mean        ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  0 ║ dbsp::trace::persistent::trace::new                  ║        330 ║         — ║    1.20s ║      3.64ms ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  1 ║ dbsp::trace::persistent::trace::add_batch_to_cf      ║      12151 ║         — ║   11.20s ║    921.53µs ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  6 ║ dbsp::trace::persistent::cursor::step_key            ║    4978267 ║         — ║    3.28s ║ 657.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  7 ║ dbsp::trace::persistent::cursor::seek_key            ║   12746181 ║         — ║   14.62s ║   1.15µs :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  2 ║ dbsp::trace::persistent::trace::rocksdb_concat_merge ║   11734064 ║      7, 6 ║   10.46s ║ 890.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  3 ║ dbsp::trace::persistent::cursor::fold_times          ║   16811499 ║         — ║ 219.70ms ║  13.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║ 10 ║ dbsp::trace::persistent::cursor::seek_val            ║     864751 ║         — ║  17.18ms ║  19.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  4 ║ dbsp::trace::persistent::cursor::step_val            ║   22821987 ║        10 ║ 265.90ms ║  11.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  5 ║ dbsp::trace::persistent::cursor::rewind_vals         ║    8980382 ║         — ║ 122.12ms ║  13.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  8 ║ dbsp::trace::persistent::cursor::fast_forward_vals   ║     769256 ║         — ║   6.13ms ║   7.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬═══════════╬══════════╬═════════════╣
║  9 ║ dbsp::trace::persistent::cursor::fold_times_through  ║    7905379 ║         — ║  93.42ms ║  11.00ns :warning:  ║
╚════╩══════════════════════════════════════════════════════╩════════════╩═══════════╩══════════╩═════════════╝


Gerd Zellweger
  [7 days ago](https://feldera.slack.com/archives/C05R2HWEDBQ/p1697226904160179?thread_ts=1697226883.303539&cid=C05R2HWEDBQ)
DBSP_KEY_VALUE as rocksdb key branch:
┌───────┬───────────┬───────┬───────────┬─────────────────┬──────────────────┬───────────────┬───────────────┬─────────────┬────────────┬────────────────┬─────────────┬─────────────┐
│ Query │ #Events   │ Cores │ Elapsed   │ Cores * Elapsed │ Throughput/Cores │ Total Usr CPU │ Total Sys CPU │ Current RSS │ Peak RSS   │ Current Commit │ Peak Commit │ Page Faults │
├───────┼───────────┼───────┼───────────┼─────────────────┼──────────────────┼───────────────┼───────────────┼─────────────┼────────────┼────────────────┼─────────────┼─────────────┤
│ q0    │ 3,000,000 │ 1     │ 604.877ms │ 604.877ms       │ 4959.685 K/s     │ 3.419s        │ 48.000ms      │ 141.88 MiB  │ 141.94 MiB │ 184.73 MiB     │ 184.83 MiB  │ 2           │
│ q1    │ 3,000,000 │ 1     │ 616.634ms │ 616.634ms       │ 4865.120 K/s     │ 3.570s        │ 39.000ms      │ 167.42 MiB  │ 167.52 MiB │ 24.67 MiB      │ 24.83 MiB   │ 0           │
│ q2    │ 3,000,000 │ 1     │ 606.644ms │ 606.644ms       │ 4945.240 K/s     │ 3.462s        │ 44.000ms      │ 184.02 MiB  │ 184.11 MiB │ 688.00 KiB     │ 848.00 KiB  │ 0           │
│ q3    │ 3,000,000 │ 1     │ 781.645ms │ 781.645ms       │ 3838.058 K/s     │ 3.736s        │ 64.000ms      │ 204.02 MiB  │ 204.09 MiB │ 240.00 KiB     │ 336.00 KiB  │ 0           │
│ q4    │ 3,000,000 │ 1     │ 7.067s    │ 7.067s          │ 424.482 K/s      │ 11.070s       │ 261.000ms     │ 581.95 MiB  │ 582.03 MiB │ 256.00 KiB     │ 352.00 KiB  │ 0           │
│ q5    │ 3,000,000 │ 1     │ 627.846ms │ 627.846ms       │ 4778.243 K/s     │ 4.139s        │ 69.000ms      │ 591.97 MiB  │ 592.05 MiB │ 224.00 KiB     │ 320.00 KiB  │ 0           │
│ q6    │ 3,000,000 │ 1     │ 7.548s    │ 7.548s          │ 397.479 K/s      │ 11.566s       │ 273.000ms     │ 680.39 MiB  │ 680.47 MiB │ 8.22 MiB       │ 8.31 MiB    │ 0           │
│ q7    │ 3,000,000 │ 1     │ 1.935s    │ 1.935s          │ 1550.562 K/s     │ 6.226s        │ 252.000ms     │ 789.27 MiB  │ 789.34 MiB │ 4.25 MiB       │ 4.34 MiB    │ 0           │
│ q8    │ 3,000,000 │ 1     │ 625.348ms │ 625.348ms       │ 4797.325 K/s     │ 3.651s        │ 61.000ms      │ 790.38 MiB  │ 790.52 MiB │ 240.00 KiB     │ 336.00 KiB  │ 0           │
│ q9    │ 3,000,000 │ 1     │ 23.598s   │ 23.598s         │ 127.129 K/s      │ 36.293s       │ 1.915s        │ 1.47 GiB    │ 1.47 GiB   │ 296.23 MiB     │ 296.33 MiB  │ 0           │
│ q12   │ 3,000,000 │ 1     │ 1.322s    │ 1.322s          │ 2268.942 K/s     │ 4.318s        │ 71.000ms      │ 1.47 GiB    │ 1.47 GiB   │ 240.00 KiB     │ 336.00 KiB  │ 0           │
│ q13   │ 3,000,000 │ 1     │ 18.675s   │ 18.675s         │ 160.639 K/s      │ 22.164s       │ 325.000ms     │ 1.47 GiB    │ 1.47 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q14   │ 3,000,000 │ 1     │ 606.701ms │ 606.701ms       │ 4944.773 K/s     │ 3.550s        │ 46.000ms      │ 1.47 GiB    │ 1.47 GiB   │ 16.00 EiB      │ 128.00 KiB  │ 0           │
│ q15   │ 3,000,000 │ 1     │ 2.756s    │ 2.756s          │ 1088.488 K/s     │ 5.626s        │ 115.000ms     │ 1.49 GiB    │ 1.49 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q16   │ 3,000,000 │ 1     │ 49.593s   │ 49.593s         │ 60.493 K/s       │ 51.895s       │ 641.000ms     │ 1.62 GiB    │ 1.62 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q17   │ 3,000,000 │ 1     │ 32.506s   │ 32.506s         │ 92.289 K/s       │ 36.006s       │ 463.000ms     │ 1.93 GiB    │ 1.93 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q18   │ 3,000,000 │ 1     │ 31.006s   │ 31.006s         │ 96.757 K/s       │ 38.385s       │ 818.000ms     │ 2.23 GiB    │ 2.23 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q19   │ 3,000,000 │ 1     │ 12.017s   │ 12.017s         │ 249.655 K/s      │ 19.537s       │ 784.000ms     │ 2.37 GiB    │ 2.37 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q20   │ 3,000,000 │ 1     │ 12.034s   │ 12.034s         │ 249.298 K/s      │ 19.081s       │ 679.000ms     │ 2.41 GiB    │ 2.41 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q21   │ 3,000,000 │ 1     │ 607.097ms │ 607.097ms       │ 4941.549 K/s     │ 3.532s        │ 51.000ms      │ 2.41 GiB    │ 2.41 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
│ q22   │ 3,000,000 │ 1     │ 609.571ms │ 609.571ms       │ 4921.492 K/s     │ 3.509s        │ 50.000ms      │ 2.41 GiB    │ 2.41 GiB   │ 16.00 EiB      │ 144.00 KiB  │ 0           │
└───────┴───────────┴───────┴───────────┴─────────────────┴──────────────────┴───────────────┴───────────────┴─────────────┴────────────┴────────────────┴─────────────┴─────────────┘
╔════╦══════════════════════════════════════════════════════╦════════════╦════════════════╦══════════╦═════════════╗
║ id ║ span name                                            ║ call count ║ called by      ║ total    ║ mean        ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  0 ║ dbsp::trace::persistent::trace::new                  ║        330 ║              — ║    6.57s ║     19.89ms ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  1 ║ dbsp::trace::persistent::trace::add_batch_to_cf      ║      12151 ║              — ║   24.89s ║      2.05ms ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║ 10 ║ dbsp::trace::persistent::cursor::seek_val            ║     873253 ║              — ║    4.00s ║      4.58µs ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  4 ║ dbsp::trace::persistent::cursor::step_val            ║   22832134 ║             10 ║   91.75s ║      4.02µs ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  8 ║ dbsp::trace::persistent::cursor::fast_forward_vals   ║     769697 ║              — ║   13.06s ║     16.97µs ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  6 ║ dbsp::trace::persistent::cursor::rewind_vals         ║    9755049 ║              8 ║   20.66s ║      2.12µs ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  5 ║ dbsp::trace::persistent::cursor::seek_key            ║   22536693 ║              6 ║   44.76s ║      1.99µs ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  7 ║ dbsp::trace::persistent::cursor::step_key            ║    4995245 ║              — ║    9.05s ║      1.81µs ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  2 ║ dbsp::trace::persistent::trace::rocksdb_concat_merge ║   79250439 ║ 10, 8, 7, 5, 4 ║    8.44s ║ 105.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  3 ║ dbsp::trace::persistent::cursor::fold_times          ║   16818683 ║              — ║ 218.35ms ║  12.00ns :warning:  ║
╠════╬══════════════════════════════════════════════════════╬════════════╬════════════════╬══════════╬═════════════╣
║  9 ║ dbsp::trace::persistent::cursor::fold_times_through  ║    7907083 ║              — ║  63.59ms ║   8.00ns :warning:  ║
╚════╩══════════════════════════════════════════════════════╩════════════╩════════════════╩══════════╩═════════════╝


- add_batch_to_cf: takes 10s less on main (I guess inserting less keys is better)
- step_val: 91.75s vs. 260ms... deserializing all values (for a given key) and keeping them as a vector in dram (instead of seeking them one by one in rocksdb) looks like the big win... Maybe also having to deserialize the key every time when we step to the next value is adding some extra overhead here too (I wasn't able to solve the alignment issue that would lead to no-cost deserialization which could potentially also skip the key).. but given the time differences there is probably more going on here it's definitely not as cheap as indexing an array which is what main does here..
- fast_forward_vals 13s vs 6ms.. (this has to do a seek)
- rewind_vals 120ms vs. 20.66s
- seek_key 44.76s vs. 14.64 .. we have a lot more rocksdb keys now because every value ends up being a key, so I suspect that's why seek time increases here because the impls of these two haven't really changed that much
- rocksdb_concat_merge: this is my horribly inefficient function with quadratic complexity (but it doesnt seem to be the bottleneck) (edited)
